### ISSUE TITLE
Different gradients with batch norm on GPU and CPU

### ISSUE THREAD

**Issue by nikonikolov** [pytorch/pytorch#34546]
## 🐛 Bug

Significantly different gradients with `torch.nn.BatchNorm1d` and `torch.nn.Linear` on GPU and CPU. 

## To Reproduce

Script to reproduce
```python
import numpy as np
import torch
import torch.nn as nn

torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

print(torch.__version__)
print(torch.version.cuda)
print(torch.backends.cudnn.version())

def run_all():
    lbns = []
    lnbns = []
    convs = []

    for device in ['cpu', 'cuda']:
        torch.manual_seed(42)
        x_linear_in = torch.randn(2, 1024).to(device)
        x_conv_in = torch.randn(2, 32, 64, 64).to(device)
        y = torch.randn(2, ).to(device)
        lbns.append(run_model(x_linear_in, y, linear_model(device, batch_norm=True), "LINEAR WITH BATCH NORM", device))
        lnbns.append(run_model(x_linear_in, y, linear_model(device, batch_norm=False), "LINEAR NO BATCH NORM", device))
        convs.append(run_model(x_conv_in, y, conv_model(device), "CONV", device))

    # Sanity checks
    print_result(lbns[0], lbns[1], "LINEAR WITH BATCH NORM")
    print_result(lnbns[0], lnbns[1], "LINEAR NO BATCH NORM")
    print_result(convs[0], convs[1], "CONV")

def print_result(values_1, values_2, name):
    print(f"\nRESULT FOR: {name}\n")
    max_diff = -np.inf
    for v1, v2 in zip(values_1, values_2):
        print(v1, v2)
        max_diff = max(abs(v1 - v2), max_diff)
    print(f"\nMAX DIFF FOR {name}: {max_diff}\n")

def linear_model(device, batch_norm):
    layer_sizes = [1024, 512, 256, 128, 64, 32, 16, 8]
    layers = [
        nn.Sequential(nn.Linear(f_in, f_out), nn.BatchNorm1d(f_out) if batch_norm else nn.Identity(), nn.LeakyReLU())
        for f_in, f_out in zip(layer_sizes[:-1], layer_sizes[1:])
    ]
    layers += [nn.Linear(layer_sizes[-1], 1)]
    model = torch.nn.Sequential(*layers).to(device)
    return model

def conv_model(device):
    layers = [
        nn.Sequential(nn.Conv2d(32, 32, kernel_size=5, padding=2), nn.BatchNorm2d(32), nn.LeakyReLU())
        for i in range(10)
    ]
    layers += [nn.Flatten(), nn.Linear(32*64*64, 1)]
    model = torch.nn.Sequential(*layers).to(device)
    return model

def run_model(x, y, model, name, device):
    y_hat = model(x)
    loss = torch.mean((y - y_hat)**2)
    loss.backward()
    print(f"{name} on {device} loss: {loss.sum()}")
    return np.array([torch.abs(p.grad).sum().cpu().numpy() for p in model.parameters()])

run_all()

```

Output:
```
1.4.0
10.1
7603
LINEAR WITH BATCH NORM on cpu loss: 0.46710404753685
LINEAR NO BATCH NORM on cpu loss: 0.7761200666427612
CONV on cpu loss: 2.450071334838867
LINEAR WITH BATCH NORM on cuda loss: 0.46710386872291565
LINEAR NO BATCH NORM on cuda loss: 0.7761200666427612
CONV on cuda loss: 2.450077533721924

RESULT FOR: LINEAR WITH BATCH NORM

1054.5361 1034.3702
5.6175622e-06 2.3710552e-06
1.7143956 1.6817329
1.6870903 1.6549473
190.00961 186.39351
5.7916884e-07 7.875437e-07
0.12345238 0.12127322
0.123077355 0.12090216
5.957692 5.8600616
1.550708e-08 2.6453515e-08
0.022549428 0.022221267
0.022247553 0.02192361
0.7192713 0.7082452
6.13243e-08 1.5854486e-07
0.029166669 0.028698694
0.028666768 0.028206829
0.65573823 0.6454811
1.7313099e-08 3.048801e-08                                                                                                                                                                                       
0.042993236 0.04236982                                                                                                                                                                                           
0.04229156 0.04167823                                                                                                                                                                                            
0.59108526 0.58249503                                                                                                                                                                                            
2.5913295e-08 1.3247777e-06                                                                                                                                                                                      
0.09252064 0.09111691                                                                                                                                                                                            
0.09095797 0.08957807                                                                                                                                                                                            
0.7392909 0.7280751                                                                                                                                                                                              
2.0057269e-07 6.462081e-08                                                                                                                                                                                       
0.19017482 0.19009309                                                                                                                                                                                            
0.22592424 0.22583872                                                                                                                                                                                            
0.90315515 0.9028712                                                                                                                                                                                             
0.8967905 0.8967482                                                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                                             
MAX DIFF FOR LINEAR WITH BATCH NORM: 20.1658935546875                                                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                                             
RESULT FOR: LINEAR NO BATCH NORM                                                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                                             
53.660133 53.660133                                                                                                                                                                                                                                                                                                          
0.06936559 0.06936559                                                                                                                                                                                                                                                                                                        
14.715334 14.715334                                                                                                                                                                                                                                                                                                          
0.12003719 0.1200372                                                                                                                                                                                                                                                                                                         
6.1398 6.139799                                                                                                                                                                                                                                                                                                              
0.24683239 0.24683236                                                                                                                                                                                                                                                                                                        
2.8269532 2.826953                                                                                                                                                                                                                                                                                                           
0.5003878 0.5003878                                                                                                                                                                                                                                                                                                          
1.6681464 1.6681461                                                                                                                                                                                                                                                                                                          
0.7146226 0.7146225                                                                                                                                                                                                                                                                                                          
1.3697281 1.3697282                                                                                                                                                                                                                                                                                                          
1.1652821 1.165282                                                                                                                                                                                                                                                                                                           
1.7599818 1.7599816                                                                                                                                                                                                                                                                                                          
2.0216236 2.0216236                                                                                                                                                                                                                                                                                                          
1.2722092 1.2722092                                                                                                                                                                                                                                                                                                          
1.6743914 1.6743914                                                                                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                                                                                             
MAX DIFF FOR LINEAR NO BATCH NORM: 9.5367431640625e-07                                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                                             
RESULT FOR: CONV                                                                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                                             
28466.773 28466.324
2.7193611e-05 2.6732683e-05
13.120243 13.11767
13.416933 13.418464
24155.852 24156.2
5.095585e-05 2.8133392e-05
11.420215 11.414161
10.268715 10.25854
20031.248 20031.934
4.109287e-05 2.1994114e-05
13.881267 13.882526
10.119639 10.13705
16632.014 16631.625
2.782326e-05 1.8000603e-05
9.418229 9.413584
8.935509 8.936967
13603.243 13603.291
1.8857118e-05 1.4536083e-05
11.498397 11.501751
9.668414 9.678661
11487.623 11487.674
1.830127e-05 1.2524426e-05
6.246866 6.2399797
4.8961506 4.8945017
9686.762 9686.385
2.1484651e-05 1.2502074e-05
5.757124 5.7587204
4.9328284 4.934167
8009.6606 8009.9004
1.0743662e-05 9.775162e-06
5.4138613 5.413865
4.328943 4.3261166
6711.6797 6711.6875
1.3400867e-05 7.458031e-06
3.9421966 3.9422052
3.7487607 3.7487648
5665.9883 5665.995
7.935356e-06 1.2770295e-05
5.5446377 5.544648
5.5413284 5.541335
160230.3 160230.52
3.0745316 3.0745354

MAX DIFF FOR CONV: 0.685546875
```

## Expected behavior

I would expect that gradients in the network, no matter the architecture, to be roughly equal across CPU and GPU. I can accept the differences in gradients for the conv network cause I can believe it's due to accumulated floating point errors across different devices. However, the error that we get with batch norm and linear layers is too much
```
MAX DIFF FOR LINEAR WITH BATCH NORM: 20.1658935546875         
```
The fact that the error goes down to the order of 1e-06 when we remove batch norm tells me there is something wrong with batch norm. Furthermore, replacing `BatchNorm1d` with `LayerNorm` ends up producing sensible errors, roughly on the order of 1e-03.

## Environment

```
Collecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.10.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti                                                                                                                                                                                                                                                                     
Nvidia driver version: 440.59                                                                                                                                                                                                                                                                                                
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.4.2                                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                                                             
Versions of relevant libraries:                                                                                                                                                                                                                                                                                              
[pip] Could not collect            
[conda] Could not collect
```
## Additional context

It's very strange that this problem is observed only with batch norm and linear layers but does not happen with conv layers. It is also possible that this is a global problem with batch norm, but we just don't observe it for this particular conv network (maybe a deeper one with different initialization might end up producing a similar problem).

I ran the above script on a GTX 1080 Ti and a RTX 2080 Ti, and the linear with batch norm error was big  in both cases. However, the error across the two different GPUs was small. Maybe there is something wrong with batch norm on CPU ...

**Comment by albanD** [comment]
Hi,

I am bit confused by your test code:
- Generating random numbers on different devices will give you different random value, even if you set the same random seed as different pseudo random number generators are used for different devices.
- It is expected that running the same code on different devices may not give you bit-wise same result. So when you stack a few of them, it is expected that you get a larger error (depending on which modules and how many you stack). Can you run the same test with a single module at a time?

**Comment by nikonikolov** [comment]
Hey, thanks for the quick response.

1. Regarding random numbers: All tensors are generated on the default device (CPU) and only then moved to the target device. Also if you check the losses, you will see they are very similar up to an admissible floating point error. You can also double check the inputs (I already did it myself - they are the same).

2. I am not asking for the same bit-wise result. I am asking for a result which is similar up to an admissible floating point error (at most up to 1.0).

I don't think the error happens to appear when I run only a single layer (as you can see, the final layers of the network get almost the same gradients both on CPU and on GPU). The problem occurs when the network is deeper. In practice we end up running deep networks, so I don't think this is unreasonable. Also sure, when we stack more layers together we get larger errors, but in the end mathematics is a deterministic thing and an error > 20.0 is unacceptable. I also don't think the that argument about 'stacking more layers leads to bigger errors' is applicable here: the errors are < 1.0 both in the case when we have a conv + batch norm network with similar depth and number of parameters or if we have a linear + layer norm network. This is a clear bug with batch norm.

**Comment by albanD** [comment]
1. Yes I agree with you

> mathematics is a deterministic thing

Yes, but not floating point operations. They specify a given precision and any value close enough to the truth is valid.

> This is a clear bug with batch norm.

Or maybe batchnorm increase the errors a lot (more than a simple Linear layer with weights close to 1).
Can you see an error in the value computed by a single batchnorm?

**Comment by nikonikolov** [comment]
> Can you see an error in the value computed by a single batchnorm?

Error is < 1.0.

> Or maybe batchnorm increase the errors a lot (more than a simple Linear layer with weights close to 1).

Sure it might increase errors, but > 20.0 is unacceptable. This is too much for a floating point error ...
1. Why does it not happen when we switch batch norm with layer norm in the network with linear layers?  The principle of batch norm and layer norm is very similar ...
2. Why does it not happen in the network with batch norm and conv layers, which has more parameters and is deeper?
3. Why the error > 20.0 when I compare results on CPU vs GPU, but is < 1.0 when I compare results on 1080 Ti and 2080 Ti?

**Comment by albanD** [comment]
> Error is < 1.0.

It should be much lower than that. <1e-5 at least.

By the way to do tests, you can generate float numbers that have 1 bit difference only by doing:
```python
import torch
x = torch.rand(10, 1024)
y = torch.rand(10, 1024)
z = torch.rand(10, 1024)

a = (x + y) + z
b = x + (y + z)
# Make a and b that have a 1 bit difference
print("(a-b).abs().max(): ", (a-b).abs().max())
```

**Comment by nikonikolov** [comment]
Yeah, it's < 1e-5. Sorry, the number I have was too relative. Anyway, can you investigate the rest of the errors:

> Sure it might increase errors, but > 20.0 is unacceptable. This is too much for a floating point error ...
> 
> 1. Why does it not happen when we switch batch norm with layer norm in the network with linear layers?  The principle of batch norm and layer norm is very similar ...
> 2. Why does it not happen in the network with batch norm and conv layers, which has more parameters and is deeper?
> 3. Why the error > 20.0 when I compare results on CPU vs GPU, but is < 1.0 when I compare results on 1080 Ti and 2080 Ti?

**Comment by albanD** [comment]
I just saw that your `run_model` actually take the absolute value and sum up all the gradients before comparing them. That could make you loose a lot of precision as well.
Can you try the same thing without summing? Just give `p.grad` back. And compare them by doing `max_diff = max((v1 - v2).abs().max(), max_diff)`.

**Comment by nikonikolov** [comment]
The last suggestion produces errors < 1e-2, which is acceptable
